lib/serviceDiscovery/engines/zookeeper/index.js

```diff
@@ -118,30 +122,81 @@ function ZooKeeperService(name, type, options) {
 	// node data is stored apart
 	this.services = {};
 
-	// connect
+	this.connectTimer = null;
+	this.clearCredentialsTimer = null;
 
-	logger.verbose('ZooKeeper client connecting to', options.hosts);
+	// Initial setup
+	this.reset();
+}
 
-	var that = this;
+util.inherits(ZooKeeperService, Service);
+
+
+ZooKeeperService.prototype.reset = function () {
+	const { options } = this;
+
+	this.client = createZooKeeperClient(options);
+
+	// After reading the code below, you might feel a bit
+	// disgusted; this is perfectly normal, but currently
+	// unavoidable.
+	//
+	// The client library we are currently using will simply
+	// cycle forever on all zookeeper node if the remote server
+	// respond with an invalid session error. While this possibly make
+	// sense for larger deployments, it also means that restarting a single
+	// ZooKeeper node would mean needing to restart the MAGE cluster.
+	//
+	// To solve this, we simply replace the client if we cannot achiever
+	// a connection for too long.
+	//
+	// See: https://github.com/mage/mage/issues/108
+	this.client.on('connected', function () {
+		if (this.connectTimer) {
+			clearTimeout(this.connectTimer);
+		}
+
+		if (this.clearCredentialsTimer) {
+			clearTimeout(this.clearCredentialsTimer);
+		}
+	});
+
+	this.client.on('disconnected', () => {
+		this.clearCredentialsTimer = setTimeout(() => {
+			var state = this.client.getState();
+
+			if (state !== zooKeeper.State.SYNC_CONNECTED) {
+				logger.verbose('Connection appears unstable, re-setting');
+				this.client.close();
+				this.reset();
+			}
+		}, CONNECT_TIMEOUT);
+	});
+
+	if (!this.closed) {
+		logger.verbose('ZooKeeper client connecting to', options.hosts);
+		return this.client.connect();
+	}
 
-	this.client.connect();
+	if (this.discovering !== false) {
```

Comment:
In order:

  1. this.closed === true means we must be connected and makes no assumption as to whether we either announce or discover
  2. this.discovering means we were discovering prior to the reset, and are expected to keep on discovering
  3. this.announcing means we were announcing prior to the reset, and are expected to keep on announcing

Note that close sets all three values to false: https://github.com/mage/mage/pull/113/files/b5686cfc24918b5f886cc0890bb9a84604a587a2#diff-df9e000e82d7187631d7095d0604e1a7R420